В цьому ДЗ потренуємось виконувати Z-тести для тестування гіпотез в різних задачах. Для коректної імплементації та інтерпретації тестів рекоемндую переглянути спершу документацію:

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html

### Завдання 1: Z-тест для однієї вибірки

**Мета**: Перевірити, чи середнє значення вибірки відрізняється від гіпотетичного середнього.

**Припущення**: Середній ріст студентів університету вважається 170 см. Ми хочемо перевірити, чи середній ріст випадкової вибірки студентів істотно більший. Вибірка задана в змінній `heights`.

Для виконання завдання імпортуйте метод для проведення `ztest`, виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення про середній зріст студентів на рівні значущості 0.05?

Використовуємо тут Z-тест, оскільки вибірка достатньо велика (більше 30 екземплярів) і можемо вважати, що її стандартне відхилення близьке до ст. відхилення генеральної сукупності, а значить - відоме.

In [1]:
heights = [174, 171, 175, 179, 170, 170, 179, 175, 169, 174, 169, 169, 173, 162, 163, 169, 166,
           173, 167, 164, 179, 170, 172, 164, 169, 175, 169, 174, 169, 169, 173, 162, 177]

In [2]:
from statsmodels.stats.weightstats import ztest
import numpy as np

In [ ]:
# Н0 - середній зріст студентів == 170 см
# Н1 - середній зріст студентів більше 170 см

In [4]:
mean_h = np.mean(heights)
alpha = 0.05
std_h = np.std(heights)
sample_size = len(heights)
null_mean = 170

In [5]:
ztest_Score, p_value = ztest(heights, value = null_mean, alternative='larger')
ztest_Score, p_value

(np.float64(0.8482235371587619), np.float64(0.19815674571091746))

Оскільки отримане Pvalue (0.198) більше за статистичну значущисть Alpha (0.05), то ми не можемо відкинути Н0. Тобто, можемо зробити висновок, що середній зріст студентів все ж дорівнює 170см.

### Завдання 2: Z-тест для двох незалежних вибірок

**Завдання 2.1.**

**Мета**: Перевірити, чи існує статистична різниця між середніми оцінками двох груп студентів.

**Припущення**: Група A отримала новий навчальний курс, а група B продовжувала зі стандартним курсом. Ми перевіряємо, чи є новий курс ефективнішим.

Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення щодо студентів на рівні значущості 0.05?

In [1]:
group_a_scores = [78.55, 72.25, 79.88, 75.  , 76.54, 74.99, 87.26, 77.93, 72.71,
       82.11, 71.9 , 79.04, 68.2 , 71.36, 78.98, 81.69, 78.86, 77.42,
       76.49, 70.61, 74.4 , 75.7 , 83.29, 79.72, 69.18, 79.62, 76.07,
       74.62, 81.06, 83.15, 82.66, 73.8 , 76.45, 79.66, 82.88, 75.6 ,
       77.07, 72.47, 72.02, 82.06]

group_b_scores = [81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.9 , 79.11, 75.44, 73.5 , 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.9 , 73.29, 70.99, 74.19]

In [ ]:
# Н0 - різниці між середніми оцінками двох груп немає
# Н1 - різниц між середніми оцінками двох груп є

In [3]:
np.mean(group_a_scores), np.mean(group_b_scores)

(np.float64(77.08125), np.float64(74.73875000000001))

In [4]:
ztest(group_a_scores, group_b_scores, alternative='two-sided') # ztest(group_a_scores, group_b_scores, value=0)

(np.float64(2.357354864751502), np.float64(0.01840564855252801))

Статистика двух груп оцінок становить 2.357, а p-value = 0.018. Оскільки p-value < статистичну значущисть alpha (0.05), то ми можемо відхилити Н0. Тобто можемо сказати, що середні оцінки двух груп студентів відрізняються.

**Завдання 2.2.**


**Мета**: Побачити, як впливає розмір вибірок на результат тесту.

**Завдання**: Уявіть, що з групи А ми маємо тільки перші 5 записів, а з групи В - усі. Таке могло статись, якщо ми вже маємо результати тестів учнів за попередньою програмою, а тести за новою програмою здали наразі лише 5 учнів і ми вирішили не чекати довше.
Виберіть перші 5 записів для групи А і проведіть t-test (для цього використовуємо наступний [метод](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html)). Виведіть значення p-value. Зробіть висновок, який результат ви бачите щодо стат. значущості на рівні 0.05 для цього експерименту?

Тут проводимо t-test оскільки одна з вибірок - дуже мала.

In [13]:
np.random.seed(42)
group_a_scores_small = np.random.choice(group_a_scores, 5, replace=False)
group_a_scores_small

array([70.61, 78.86, 81.69, 76.07, 76.54])

In [15]:
from scipy.stats import ttest_ind

In [16]:
ttest_ind(group_a_scores_small, group_b_scores, alternative='two-sided')

TtestResult(statistic=np.float64(0.9477312090698231), pvalue=np.float64(0.34856230273170563), df=np.float64(43.0))

Статистика двух груп оцінок становить 0.95, а p-value = 0.348. Оскільки p-value > за статистичну значущисть alpha (0.05), то ми НЕ можемо відхилити Н0. Тобто можемо сказати, що середні оцінки двух груп студентів не відрізняються.


### Завдання 3\*: Z-тест для двох **пов'язаних** вибірок

**Мета**: Перевірити, чи вплинув тренінг на продуктивність працівників, коли всі справробітники проходили тренінг.

**Припущення**: Наші працівники проходять однаковий тренінг, і ми хочемо з'ясувати, чи покращилися їхні показники після тренінгу на рівні значущості 0.05. Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи поліпшились показники працівників на рівні значущості 0.05?

Зверніть увагу, ці вибірки повʼязані між собою, значить не буде коректно виконати z-test між ними, натомість, ми можемо порівняти, чи різниця в результатах (продуктивності співробітників) є стат значущою.

In [17]:
before_training = [57.82, 37.63, 36.8 , 55.22, 52.97, 52.5 , 53.46, 43.2 , 52.32,
       52.93, 42.86, 68.66, 54.74, 38.09, 56.57, 40.25, 57.87, 61.59,
       41.79, 59.63, 54.13, 58.22, 68.97, 47.55, 42.46, 41.1 , 41.84,
       49.23, 53.41, 52.77]

after_training = [62.47, 40.66, 42.7 , 57.69, 61.41, 56.76, 54.75, 44.06, 56.29,
       55.48, 47.28, 72.6 , 57.59, 39.39, 56.54, 42.36, 62.58, 65.01,
       42.3 , 62.98, 57.9 , 59.45, 72.28, 50.66, 43.18, 44.82, 45.96,
       54.4 , 58.52, 53.01]

In [22]:
results = [round(x-y, 2) for x in before_training for y in after_training]

In [23]:
np.mean(results)

np.float64(-3.15)

In [ ]:
# Н0 - різниці в продуктивності немає, value=0
# Н2 - продуктивність зросла

In [25]:
ztest(results, value=0, alternative='larger')

(np.float64(-7.610505476949509), np.float64(0.9999999999999863))

Різниця в результатах продуктивності бспівробітників є статистично НЕ значущою, бо p-value 0.99 більше за статистичну значущисть alpha 0.05 і ми не можемо відхилити Н0.